modules/spring-config/src/main/java/org/mule/runtime/config/internal/DependencyGraphLifecycleObjectSorter.java

```diff
@@ -0,0 +1,191 @@
+      dependencyGraphs.add(graph);
+    }
+    this.lifecycleObjectNameOrderMap = new HashMap<>();
+  }
+
+  /**
+   * Building a single dependency graph(bucket) for each lifecycle type
+   * 
+   * @param beanName
+   * @param currentObject
+   */
+  @Override
+  public void addObject(String beanName, Object currentObject) {
+
+    requireNonNull(currentObject, "currentObject cannot be null");
+
+    if (stream(orderedLifecycleTypes).noneMatch(x -> x.isInstance(currentObject))) {
+      return;
+    }
+
+    DefaultDirectedGraph<BeanWrapper, DefaultEdge> dependencyGraph = getDependencyGraphForLifecycleType(currentObject);
+    CycleDetector cycleDetector = new CycleDetector(dependencyGraph);
+
+    BeanWrapper currentVertex = new BeanWrapper(beanName, currentObject);
+    dependencyGraph.addVertex(currentVertex);
+
+    // get (direct) prerequisite objects for the current object
+    // List<Pair<String, Object>> prerequisiteObjects = resolver.getDirectBeanDependencies(beanName);
+    Map<BeanWrapper, List<BeanWrapper>> prerequisiteObjectsMap =
+        resolver.getTransitiveDependencies(beanName, getDependencyGraphIndex(currentObject));
+
+    // add direct prerequisites to the graph & create edges(current object -> prerequisite)
+    for (BeanWrapper source : prerequisiteObjectsMap.keySet()) {
+      List<BeanWrapper> prerequisiteObjects = prerequisiteObjectsMap.get(source);
+      BeanWrapper current = new BeanWrapper(source.getName(), source.getWrappedObject());
+      dependencyGraph.addVertex(current);
+      if (prerequisiteObjects.isEmpty()) {
+        continue;
+      }
+      prerequisiteObjects.forEach(
+                                  prerequisite -> {
+                                    String preReqName = prerequisite.getName();
+                                    Object preReqObject = prerequisite.getWrappedObject();
+
+                                    BeanWrapper preReqVertex = new BeanWrapper(preReqName, preReqObject);
+                                    dependencyGraph.addVertex(preReqVertex);
+
+                                    // remove any additional edge that creates cycle
+                                    if (!dependencyGraph.containsEdge(preReqVertex, current)) {
+                                      dependencyGraph.addEdge(current, preReqVertex);
+                                      if (cycleDetector.detectCycles()) {
+                                        dependencyGraph.removeEdge(current, preReqVertex);
+                                      }
+                                    }
+                                  });
+    }
+  }
+
+  /**
+   * Provides the index of the graph(bucket) the current object should be added to
+   * 
+   * @param currentObject
+   * @return
+   */
+  private int getDependencyGraphIndex(Object currentObject) {
+    Class<?> clazz = stream(orderedLifecycleTypes).filter(x -> x.isInstance(currentObject)).findFirst().get();
```

Comment:
This method is walking through the orderedLifecycleTypes three times: stream() + asList() + indexOf().
Given you're working on performance, I think it would be preferable to write a single for loop.

If we decide to let it as is, please ensure the Optional returned by findFirst() is present before using get(), or use orElse() instead

Reply:
Replaced it with a single for loop :)